File: turtle-soup-bot/src/main/kotlin/io/github/kapu/turtlesoup/mq/handler/GameCommandHandler.kt
package io.github.kapu.turtlesoup.mq.handler

import io.github.kapu.turtlesoup.bridge.SurrenderHandler
import io.github.kapu.turtlesoup.config.GameConstants
import io.github.kapu.turtlesoup.config.PuzzleConstants
import io.github.kapu.turtlesoup.models.Command
import io.github.kapu.turtlesoup.models.GameState
import io.github.kapu.turtlesoup.mq.MessageBuilder
import io.github.kapu.turtlesoup.mq.ValkeyMQReplyPublisher
import io.github.kapu.turtlesoup.mq.models.InboundMessage
import io.github.kapu.turtlesoup.mq.models.OutboundMessage
import io.github.kapu.turtlesoup.rest.LlmRestClient
import io.github.kapu.turtlesoup.service.GameService
import io.github.kapu.turtlesoup.utils.GameAlreadyStartedException
import io.github.kapu.turtlesoup.utils.MessageKeys
import io.github.kapu.turtlesoup.utils.MessageProvider
import io.github.oshai.kotlinlogging.KotlinLogging

/** 게임 명령어 처리 핸들러 */
@Suppress("TooManyFunctions")
class GameCommandHandler(
    private val gameService: GameService,
    private val surrenderHandler: SurrenderHandler,
    private val messageProvider: MessageProvider,
    private val publisher: ValkeyMQReplyPublisher,
    private val restClient: LlmRestClient,
) {
    private val messageBuilder = MessageBuilder(messageProvider)

    suspend fun processCommand(
        message: InboundMessage,
        command: Command,
    ): String {
        if (shouldRegisterPlayer(command)) {
            gameService.registerPlayer(message.chatId, message.userId)
        }

        val handlers: Map<Class<out Command>, suspend () -> String> =
            mapOf(
                Command.Hint::class.java to { handleHint(message) },
                Command.Problem::class.java to { handleProblem(message) },
                Command.Surrender::class.java to {
                    surrenderHandler.handleConsensus(message.chatId, message.userId)
                },
                Command.Agree::class.java to { surrenderHandler.handleAgree(message.chatId, message.userId) },
                Command.Summary::class.java to { handleSummary(message) },
                Command.Help::class.java to { messageProvider.get(MessageKeys.HELP_MESSAGE) },
                Command.Unknown::class.java to { messageProvider.get(MessageKeys.ERROR_UNKNOWN_COMMAND) },
            )

        return when (command) {
            is Command.Start -> handleStart(message, command)
            is Command.Ask -> handleAsk(message, command.question)
            is Command.Answer -> handleAnswer(message, command.answer)
            else ->
                handlers[command::class.java]?.invoke()
                    ?: messageProvider.get(MessageKeys.ERROR_UNKNOWN_COMMAND)
        }
    }

    private suspend fun handleStart(
        message: InboundMessage,
        command: Command.Start,
    ): String {
        val difficulty = resolveDifficulty(command)
        val startState = startOrResumeGame(message, difficulty.value)

        publishScenario(message, startState.state, startState.isResuming)

        val instructionMessage = buildInstructionMessage(startState.state, startState.isResuming)
        return difficulty.warning?.let { "$it\n\n$instructionMessage" } ?: instructionMessage
    }

    private suspend fun handleAsk(
        message: InboundMessage,
        question: String,
    ): String {
        logger.info { "handleAsk_start session_id=${message.chatId} question=$question" }
        val (state, result) = gameService.askQuestion(sessionId = message.chatId, question = question)
        logger.info { "handleAsk_complete session_id=${message.chatId}" }
        return messageProvider.get(MessageKeys.ANSWER_RESPONSE_SINGLE, "answer" to result.answer)
    }

    private suspend fun handleAnswer(
        message: InboundMessage,
        answer: String,
    ): String {
        val result = gameService.submitAnswer(sessionId = message.chatId, answer = answer)

        return when {
            result.isCorrect ->
                messageProvider.get(
                    MessageKeys.ANSWER_CORRECT,
                    "explanation" to result.explanation,
                    "questionCount" to result.questionCount.toString(),
                    "hintCount" to result.hintCount.toString(),
                    "maxHints" to result.maxHints.toString(),
                    "hintBlock" to messageBuilder.buildHintBlock(result.hintsUsed),
                )
            result.isClose -> messageProvider.get(MessageKeys.ANSWER_CLOSE_CALL)
            else -> messageProvider.get(MessageKeys.ANSWER_INCORRECT)
        }
    }

    private suspend fun handleHint(message: InboundMessage): String {
        val (state, hint) = gameService.requestHint(sessionId = message.chatId)
        return messageProvider.get(
            MessageKeys.HINT_GENERATED,
            "hintNumber" to state.hintsUsed.toString(),
            "content" to hint,
        )
    }

    private suspend fun handleProblem(message: InboundMessage): String {
        val state = gameService.getGameState(sessionId = message.chatId)
        val scenario = state.puzzle?.scenario ?: messageProvider.get(MessageKeys.FALLBACK_PUZZLE_NOT_FOUND)
        return messageProvider.get(
            MessageKeys.PROBLEM_DISPLAY,
            "scenario" to scenario,
            "questionCount" to state.questionCount.toString(),
            "hintCount" to state.hintsUsed.toString(),
            "maxHints" to GameConstants.MAX_HINTS.toString(),
        )
    }

    private suspend fun handleSummary(message: InboundMessage): String {
        val state = gameService.getGameState(sessionId = message.chatId)
        return messageBuilder.buildSummary(state.history)
    }

    companion object {
        private val logger = KotlinLogging.logger {}
    }

    private fun resolveDifficulty(command: Command.Start): DifficultySelection {
        if (command.hasInvalidInput) {
            return DifficultySelection(
                value = null,
                warning =
                    messageProvider.get(
                        MessageKeys.START_INVALID_DIFFICULTY,
                        "min" to PuzzleConstants.MIN_DIFFICULTY,
                        "max" to PuzzleConstants.MAX_DIFFICULTY,
                    ),
            )
        }

        val desired = command.difficulty ?: return DifficultySelection(null, null)
        return if (desired in PuzzleConstants.MIN_DIFFICULTY..PuzzleConstants.MAX_DIFFICULTY) {
            DifficultySelection(desired, null)
        } else {
            DifficultySelection(
                value = null,
                warning =
                    messageProvider.get(
                        MessageKeys.START_INVALID_DIFFICULTY,
                        "min" to PuzzleConstants.MIN_DIFFICULTY,
                        "max" to PuzzleConstants.MAX_DIFFICULTY,
                    ),
            )
        }
    }

    private suspend fun startOrResumeGame(
        message: InboundMessage,
        difficulty: Int?,
    ): StartState {
        return runCatching {
            val state =
                gameService.startGame(
                    sessionId = message.chatId,
                    userId = message.userId,
                    chatId = message.chatId,
                    difficulty = difficulty,
                    category = null,
                    theme = null,
                )
            StartState(state = state, isResuming = false)
        }.getOrElse { error ->
            if (error is GameAlreadyStartedException) {
                logger.debug { "game_already_started session_id=${message.chatId} resuming" }
                StartState(state = gameService.getGameState(message.chatId), isResuming = true)
            } else {
                throw error
            }
        }
    }

    private suspend fun publishScenario(
        message: InboundMessage,
        state: GameState,
        isResuming: Boolean,
    ) {
        val puzzle = state.puzzle
        val scenario = puzzle?.scenario ?: messageProvider.get(MessageKeys.FALLBACK_PUZZLE_NOT_FOUND)
        val difficulty = puzzle?.difficulty ?: PuzzleConstants.DEFAULT_DIFFICULTY
        val scenarioMessage =
            if (isResuming) {
                messageProvider.get(MessageKeys.START_RESUME, "scenario" to scenario)
            } else {
                messageProvider.get(
                    MessageKeys.START_SCENARIO,
                    "scenario" to scenario,
                    "difficulty" to buildDifficultyStars(difficulty),
                )
            }

        publisher.publish(
            OutboundMessage.Final(
                chatId = message.chatId,
                text = scenarioMessage,
                threadId = message.threadId,
            ),
        )
    }

    private fun buildInstructionMessage(
        state: GameState,
        isResuming: Boolean,
    ): String =
        if (isResuming) {
            messageProvider.get(
                MessageKeys.START_RESUME_STATUS,
                "questionCount" to state.questionCount,
                "hintCount" to state.hintsUsed,
            )
        } else {
            messageProvider.get(MessageKeys.START_INSTRUCTION)
        }

    private fun buildDifficultyStars(difficulty: Int): String {
        val clamped = difficulty.coerceIn(PuzzleConstants.MIN_DIFFICULTY, PuzzleConstants.MAX_DIFFICULTY)
        return "★".repeat(clamped) + "☆".repeat(PuzzleConstants.MAX_DIFFICULTY - clamped)
    }

    private fun shouldRegisterPlayer(command: Command): Boolean =
        when (command) {
            is Command.Help, is Command.Unknown -> false
            else -> true
        }
}

private data class DifficultySelection(
    val value: Int?,
    val warning: String?,
)

private data class StartState(
    val state: GameState,
    val isResuming: Boolean,
)

File: game-bot-go/internal/turtlesoup/mq/game_command_handler.go
package mq

import (
	"context"
	"errors"
	"fmt"
	"log/slog"
	"strings"

	"github.com/park285/llm-kakao-bots/game-bot-go/internal/common/messageprovider"
	"github.com/park285/llm-kakao-bots/game-bot-go/internal/common/mqmsg"
	tsconfig "github.com/park285/llm-kakao-bots/game-bot-go/internal/turtlesoup/config"
	tserrors "github.com/park285/llm-kakao-bots/game-bot-go/internal/turtlesoup/errors"
	tsmessages "github.com/park285/llm-kakao-bots/game-bot-go/internal/turtlesoup/messages"
	tsmodel "github.com/park285/llm-kakao-bots/game-bot-go/internal/turtlesoup/model"
	tssvc "github.com/park285/llm-kakao-bots/game-bot-go/internal/turtlesoup/service"
)

// GameCommandHandler: 사용자의 파싱된 명령어를 받아 실제 바다거북스프 게임 로직(Service)을 호출하고 결과를 반환합니다.
type GameCommandHandler struct {
	gameService      *tssvc.GameService
	surrenderHandler *SurrenderHandler
	msgProvider      *messageprovider.Provider
	messageBuilder   *MessageBuilder
	logger           *slog.Logger
}

// NewGameCommandHandler: 새로운 GameCommandHandler 인스턴스를 생성합니다.
func NewGameCommandHandler(
	gameService *tssvc.GameService,
	surrenderHandler *SurrenderHandler,
	msgProvider *messageprovider.Provider,
	messageBuilder *MessageBuilder,
	logger *slog.Logger,
) *GameCommandHandler {
	return &GameCommandHandler{
		gameService:      gameService,
		surrenderHandler: surrenderHandler,
		msgProvider:      msgProvider,
		messageBuilder:   messageBuilder,
		logger:           logger,
	}
}

// ProcessCommand: 명령어의 종류(Start, Ask, Answer 등)에 따라 적절한 핸들러 로직을 분기하여 실행합니다.
func (h *GameCommandHandler) ProcessCommand(ctx context.Context, message mqmsg.InboundMessage, command Command) (string, error) {
	if h.shouldRegisterPlayer(command) {
		_ = h.gameService.RegisterPlayer(ctx, message.ChatID, message.UserID)
	}

	switch command.Kind {
	case CommandStart:
		return h.handleStart(ctx, message, command)
	case CommandAsk:
		return h.handleAsk(ctx, message, command.Question)
	case CommandAnswer:
		return h.handleAnswer(ctx, message, command.Answer)
	case CommandHint:
		return h.handleHint(ctx, message)
	case CommandProblem:
		return h.handleProblem(ctx, message)
	case CommandSurrender:
		return h.surrenderHandler.HandleConsensus(ctx, message.ChatID, message.UserID)
	case CommandAgree:
		return h.surrenderHandler.HandleAgree(ctx, message.ChatID, message.UserID)
	case CommandSummary:
		return h.handleSummary(ctx, message)
	case CommandHelp:
		return h.msgProvider.Get(tsmessages.HelpMessage), nil
	case CommandUnknown:
		return h.msgProvider.Get(tsmessages.ErrorUnknownCommand), nil
	default:
		return h.msgProvider.Get(tsmessages.ErrorUnknownCommand), nil
	}
}

// handleStart: 새로운 게임을 시작하거나 기존 게임을 재개한다. 시나리오와 게임 규칙을 안내한다.
func (h *GameCommandHandler) handleStart(ctx context.Context, message mqmsg.InboundMessage, command Command) (string, error) {
	selection := h.resolveDifficulty(command)
	startState, err := h.startOrResumeGame(ctx, message, selection.Value)
	if err != nil {
		return "", fmt.Errorf("start or resume game failed: %w", err)
	}

	if startState.IsResuming {
		return h.composeStartReply(selection, startState.State, true), nil
	}

	scenario := h.buildScenarioMessage(startState.State, false)
	instruction := h.buildInstructionMessage(startState.State, false)
	if selection.Warning != "" {
		return selection.Warning + "\n\n" + scenario + "\n\n" + instruction, nil
	}
	return scenario + "\n\n" + instruction, nil
}

// handleAsk: 사용자의 질문을 AI에게 전달하여 "예/아니오" 답변을 받아 반환한다.
func (h *GameCommandHandler) handleAsk(ctx context.Context, message mqmsg.InboundMessage, question string) (string, error) {
	h.logger.Debug("handleAsk_start", "session_id", message.ChatID)
	_, result, err := h.gameService.AskQuestion(ctx, message.ChatID, question)
	if err != nil {
		return "", fmt.Errorf("ask question failed: %w", err)
	}
	h.logger.Debug("handleAsk_complete", "session_id", message.ChatID)

	return h.msgProvider.Get(tsmessages.AnswerResponseSingle, messageprovider.P("answer", result.Answer)), nil
}

// handleAnswer: 사용자가 제출한 정답을 검증하고, 결과(정답/오답/근접)에 따른 메시지를 생성한다.
func (h *GameCommandHandler) handleAnswer(ctx context.Context, message mqmsg.InboundMessage, answer string) (string, error) {
	result, err := h.gameService.SubmitAnswer(ctx, message.ChatID, answer)
	if err != nil {
		return "", fmt.Errorf("submit answer failed: %w", err)
	}

	switch {
	case result.IsCorrect():
		return h.msgProvider.Get(
			tsmessages.AnswerCorrect,
			messageprovider.P("explanation", result.Explanation),
			messageprovider.P("questionCount", result.QuestionCount),
			messageprovider.P("hintCount", result.HintCount),
			messageprovider.P("maxHints", result.MaxHints),
			messageprovider.P("hintBlock", h.messageBuilder.BuildHintBlock(result.HintsUsed)),
		), nil
	case result.IsClose():
		return h.msgProvider.Get(tsmessages.AnswerCloseCall), nil
	default:
		return h.msgProvider.Get(tsmessages.AnswerIncorrect), nil
	}
}

// handleHint: 게임 진행 중 힌트를 생성하고 제공한다. 힌트 카운트를 차감한다.
func (h *GameCommandHandler) handleHint(ctx context.Context, message mqmsg.InboundMessage) (string, error) {
	state, hint, err := h.gameService.RequestHint(ctx, message.ChatID)
	if err != nil {
		return "", fmt.Errorf("request hint failed: %w", err)
	}
	return h.msgProvider.Get(
		tsmessages.HintGenerated,
		messageprovider.P("hintNumber", state.HintsUsed),
		messageprovider.P("content", hint),
	), nil
}

// handleProblem: 현재 문제의 시나리오와 상태(남은 힌트 등)를 다시 보여준다.
func (h *GameCommandHandler) handleProblem(ctx context.Context, message mqmsg.InboundMessage) (string, error) {
	state, err := h.gameService.GetGameState(ctx, message.ChatID)
	if err != nil {
		return "", fmt.Errorf("get game state failed: %w", err)
	}

	scenario := h.msgProvider.Get(tsmessages.FallbackPuzzleNotFound)
	if state.Puzzle != nil && strings.TrimSpace(state.Puzzle.Scenario) != "" {
		scenario = state.Puzzle.Scenario
	}

	return h.msgProvider.Get(
		tsmessages.ProblemDisplay,
		messageprovider.P("scenario", scenario),
		messageprovider.P("questionCount", state.QuestionCount),
		messageprovider.P("hintCount", state.HintsUsed),
		messageprovider.P("maxHints", tsconfig.GameMaxHints),
	), nil
}

// handleSummary: 지금까지 주고받은 질문과 답변의 이력을 요약하여 보여준다.
func (h *GameCommandHandler) handleSummary(ctx context.Context, message mqmsg.InboundMessage) (string, error) {
	state, err := h.gameService.GetGameState(ctx, message.ChatID)
	if err != nil {
		return "", fmt.Errorf("get game state failed: %w", err)
	}
	return h.messageBuilder.BuildSummary(state.History), nil
}

type difficultySelection struct {
	Value   *int
	Warning string
}

func (h *GameCommandHandler) resolveDifficulty(command Command) difficultySelection {
	if command.HasInvalidInput {
		return difficultySelection{
			Value: nil,
			Warning: h.msgProvider.Get(
				tsmessages.StartInvalidDifficulty,
				messageprovider.P("min", tsconfig.PuzzleMinDifficulty),
				messageprovider.P("max", tsconfig.PuzzleMaxDifficulty),
			),
		}
	}

	if command.Difficulty == nil {
		return difficultySelection{Value: nil}
	}

	desired := *command.Difficulty
	if desired >= tsconfig.PuzzleMinDifficulty && desired <= tsconfig.PuzzleMaxDifficulty {
		return difficultySelection{Value: &desired}
	}

	return difficultySelection{
		Value: nil,
		Warning: h.msgProvider.Get(
			tsmessages.StartInvalidDifficulty,
			messageprovider.P("min", tsconfig.PuzzleMinDifficulty),
			messageprovider.P("max", tsconfig.PuzzleMaxDifficulty),
		),
	}
}

type startState struct {
	State      tsmodel.GameState
	IsResuming bool
}

func (h *GameCommandHandler) startOrResumeGame(ctx context.Context, message mqmsg.InboundMessage, difficulty *int) (startState, error) {
	state, err := h.gameService.StartGame(ctx, message.ChatID, message.UserID, message.ChatID, difficulty, nil, nil)
	if err == nil {
		return startState{State: state, IsResuming: false}, nil
	}

	var alreadyStarted *tserrors.GameAlreadyStartedError
	if errors.As(err, &alreadyStarted) {
		resumed, errLoad := h.gameService.GetGameState(ctx, message.ChatID)
		if errLoad != nil {
			return startState{}, fmt.Errorf("resume game load failed: %w", errLoad)
		}
		h.logger.Debug("game_already_started_resuming", "session_id", message.ChatID)
		return startState{State: resumed, IsResuming: true}, nil
	}

	return startState{}, fmt.Errorf("start game failed: %w", err)
}

func (h *GameCommandHandler) buildScenarioMessage(state tsmodel.GameState, isResuming bool) string {
	scenario := h.msgProvider.Get(tsmessages.FallbackPuzzleNotFound)
	difficulty := tsconfig.PuzzleDefaultDifficulty
	if state.Puzzle != nil {
		if strings.TrimSpace(state.Puzzle.Scenario) != "" {
			scenario = state.Puzzle.Scenario
		}
		difficulty = state.Puzzle.Difficulty
	}

	if isResuming {
		return h.msgProvider.Get(tsmessages.StartResume, messageprovider.P("scenario", scenario))
	}

	return h.msgProvider.Get(
		tsmessages.StartScenario,
		messageprovider.P("scenario", scenario),
		messageprovider.P("difficulty", buildDifficultyStars(difficulty)),
	)
}

func (h *GameCommandHandler) composeStartReply(selection difficultySelection, state tsmodel.GameState, isResuming bool) string {
	scenarioMessage := h.buildScenarioMessage(state, isResuming)
	instruction := h.buildInstructionMessage(state, isResuming)

	parts := make([]string, 0, 3)
	if !isResuming && selection.Warning != "" {
		parts = append(parts, selection.Warning)
	}
	parts = append(parts, scenarioMessage, instruction)
	return strings.Join(parts, "\n\n")
}

func (h *GameCommandHandler) buildInstructionMessage(state tsmodel.GameState, isResuming bool) string {
	if isResuming {
		return h.msgProvider.Get(
			tsmessages.StartResumeStatus,
			messageprovider.P("questionCount", state.QuestionCount),
			messageprovider.P("hintCount", state.HintsUsed),
		)
	}
	return h.msgProvider.Get(tsmessages.StartInstruction)
}

func buildDifficultyStars(difficulty int) string {
	clamped := difficulty
	if clamped < tsconfig.PuzzleMinDifficulty {
		clamped = tsconfig.PuzzleMinDifficulty
	}
	if clamped > tsconfig.PuzzleMaxDifficulty {
		clamped = tsconfig.PuzzleMaxDifficulty
	}

	return strings.Repeat("★", clamped) + strings.Repeat("☆", tsconfig.PuzzleMaxDifficulty-clamped)
}

func (h *GameCommandHandler) shouldRegisterPlayer(command Command) bool {
	switch command.Kind {
	case CommandHelp, CommandUnknown:
		return false
	default:
		return true
	}
}
